app/models/case.rb

```
@@ -84,11 +91,24 @@ class Case < ApplicationRecord
 
   before_validation :assign_default_subject_if_unset
 
+  before_create :set_display_id
   after_create :send_new_case_email
   after_update :maybe_send_new_assignee_email
 
   scope :active, -> { where(state: 'open') }
 
+  def to_param
+    self.display_id.parameterize.upcase
+  end
+
+  def self.find_from_id(id)
+    if /^[0-9]+$/.match(id)  # It's just a numeric ID
+      Case.find(id).decorate
+    else # It has non-digits in - let's assume it's a display ID
+      Case.find_by_display_id(id.upcase)
```

I think we should use the `find_by_display_id!` method here, as this will mean we fail fast with the usual `RecordNotFound` error if we attempt to find a non-existent Case (`Case.find` does this by default I believe so that branch is fine).

I think using the `find_by_*!` version of methods is usually what we want (unless we want to explicitly check for a `nil` record at the point we use the finder method), as that means we can either guarantee we will get an error raised or an AR object; using the non-`!` version means we can get `nil` which can then cause a `NoMethodError` in a distant place. In particular in this case this means I get
```
NoMethodError in CasesController#show
undefined method `associated_model' for nil:NilClass
```
when I navigate to a non-existent Case like `http://localhost:3000/cases/bar123`

If we do make this change I think it would also make sense to rename this method to `find_from_id!`, to match the AR convention since it is a thin wrapper around calling the correct AR method.

---

Good shout - both these changes work in concert together, so we no longer need the conditional in `assign_scope`.